core/IssueQueue.cpp

```diff
@@ -0,0 +1,228 @@
+    : sparta::Unit(node),
+      reg_file_(olympia::coreutils::determineRegisterFile(
+          node->getGroup().substr(node->getGroup().find("_") + 1))),
+      scheduler_size_(p->scheduler_size), in_order_issue_(p->in_order_issue) {
+  in_execute_inst_.registerConsumerHandler(CREATE_SPARTA_HANDLER_WITH_DATA(
+      IssueQueue, getInstsFromDispatch_, InstPtr));
+  in_exe_pipe_done_.registerConsumerHandler(
+      CREATE_SPARTA_HANDLER_WITH_DATA(IssueQueue, readyExeUnit_, uint32_t));
+  in_reorder_flush_.registerConsumerHandler(CREATE_SPARTA_HANDLER_WITH_DATA(
+      IssueQueue, flushInst_, FlushManager::FlushingCriteria));
+  sparta::StartupEvent(node,
+                       CREATE_SPARTA_HANDLER(IssueQueue, setupIssueQueue_));
+  node->getParent()
+      ->registerForNotification<bool, IssueQueue, &IssueQueue::onROBTerminate_>(
+          this, "rob_stopped_notif_channel",
+          false /* ROB maybe not be constructed yet */);
+}
+
+void IssueQueue::setupIssueQueue_() {
+  std::vector<core_types::RegFile> reg_files = {core_types::RF_INTEGER,
+                                                core_types::RF_FLOAT};
+  // if we ever move to multicore, we only want to have resources look for scoreboard in their cpu
+  // if we're running a test where we only have top.rename or top.issue_queue, then we can just use the root
+  auto cpu_node = getContainer()->findAncestorByName("core.*");
+  if(cpu_node == nullptr){
+    cpu_node = getContainer()->getRoot();
+  }
+  for (const auto rf : reg_files) {
+    scoreboard_views_[rf].reset(new sparta::ScoreboardView(
+        getContainer()->getName(), core_types::regfile_names[rf],
+        cpu_node));
+  }
+  out_scheduler_credits_.send(scheduler_size_);
+}
+
+void IssueQueue::onROBTerminate_(const bool &val) {
+  rob_stopped_simulation_ = val;
+}
+
+void IssueQueue::dumpDebugContent_(std::ostream &output) const {
+  output << "IssueQueue Contents" << std::endl;
+  for (const auto &entry : ready_queue_) {
+    output << '\t' << entry << std::endl;
+  }
+}
+
+void IssueQueue::onStartingTeardown_() {
+  if ((false == rob_stopped_simulation_) && (false == ready_queue_.empty())) {
+    dumpDebugContent_(std::cerr);
+    sparta_assert(false, "Issue queue has pending instructions");
+  }
+}
+
+void IssueQueue::setExePipe(std::string exe_pipe_name,
+                            olympia::ExecutePipe *exe_pipe) {
+  exe_pipes_[exe_pipe_name] = exe_pipe;
+}
+
+void IssueQueue::setExePipeMapping(InstArchInfo::TargetPipe tgt_pipe,
+                                   olympia::ExecutePipe *exe_pipe) {
+  if (pipe_exe_pipe_mapping_.find(tgt_pipe) == pipe_exe_pipe_mapping_.end()) {
+    pipe_exe_pipe_mapping_[tgt_pipe] = std::vector<olympia::ExecutePipe *>();
+  }
+  auto pipe_vector = pipe_exe_pipe_mapping_.find(tgt_pipe);
+  pipe_vector->second.emplace_back(exe_pipe);
+  pipes_.push_back(exe_pipe);
+}
+
+std::unordered_map<std::string, olympia::ExecutePipe *>
```

Comment:
Suggest making a type for this.

Reply:
Implemented in [fac0938](https://github.com/riscv-software-src/riscv-perf-model/pull/140/commits/fac09383254362f7866db8b5c5b51a277f968f50)